[DM-3921] s3s3 is a microservice to move s3 objects from one S3 compliant service to another.

jmatt opened the pull request:
# s3s3

s3s3 is a microservice to move files from one S3 compliant service to another (Swift, Ceph, AWS).

s3s3 has two services. `s3s3.scripts.listen` subscribes and listens to a redis pubsub channel waiting for s3 object names. When notified it calls `s3s3.api.upload` and uploads that S3 object from the source to destination(s). `s3s3.scripts.bucket` uploads all keys in one s3 bucket to another.

`s3s3.scripts.listen` is meant to be a real-time daemon.

`s3s3.scripts.bucket` is meant to run periodically as a cron job. Anything missed by `s3s3.scirpts.listen` will be handled by this service.
## Summary

s3s3 is required because Ceph and Swift are not feature complete with AWS S3. Many available libraries that work well with AWS S3 do not work with Ceph and Swift.

Some examples: `key.size` and `key.md5` do not work with Ceph S3 without fetching the contents of the key (or s3 object). Multipart uploads are not reliable with Ceph S3. V4 signatures are not supported by Ceph S3.
## Configuration

s3s3 must be configured. Configuration can be found in the `s3s3.config` module. A source, destination and pubsub section are required. Multiple destination connections are supported but there was minimal testing. To signify a section is a destination connection it must start with `'dest'`.

Example template: https://github.com/jmatt/s3s3/blob/master/s3s3/s3s3.ini.dist

``` conf
[source]
     aws_access_key_id = {YOUR_AWS_ACCESS_KEY_ID}
     aws_secret_access_key = {YOUR_AWS_SECRET_ACCESS_KEY}
     bucket_name = {YOUR_S3_BUCKET}
     host = {YOUR_CEPH_S3_ENDPOINT}
     verify_md5 = False # Verify md5 during s3 operations. 
     is_secure = True # Optional
     calling_format = OrdinaryCallingFormat # Optional
[destination]
     aws_access_key_id = {YOUR_AWS_ACCESS_KEY_ID}
     aws_secret_access_key = {YOUR_AWS_SECRET_ACCESS_KEY}
     bucket_name = {YOUR_S3_BUCKET}
     verify_md5 = False # Verify md5 during s3 operations. 
     is_secure = True # Optional
```
## Install

s3s3 requires python3 and redis. It was tested with python3.4 and python3.5. And redis 3.x.

``` bash
pip install s3s3
```

Pip install s3s3.

``` bash
s3s3listen --config /path/to/s3s3.ini
```

This will use the configuration to build source and destination boto connections, connect to redis and start listening on the `backup` channel. Any messages pushed to that channel will be considered source s3 key names and will attempt to be uploaded to the destination connection(s).

``` bash
s3s3bucket --config /path/pto/s3s3.ini
```

This will use the configuration to build source and destination boto connections and duplicate the source bucket in the destination bucket.
## Client

There are two clients. One for each service. `s3s3.client.ListenClient` is a client to listen to the `'backup'` redis pubsub channel and call `s3s3.client.on_notify`. `s3s3.client.BucketClient` is a client that uses the configuration to provide access to the duplicate_bucket API function.

`````` python
```class ListenClient(Client):
``````
## API

The API can be found in `s3s3.api` module.

``` python
def create_connection(connection_args):
```

Creates a boto connection from the `connection_args` dictionary.

``` python
def upload(source_key, dest_key, verify_md5=False):
```

Upload the source key (S3 object) to the destination key. If `verify_md5` is true then verify md5s match.

``` python
def duplicate_bucket(source_bucket, dest_bucket, verify_md5=False):
```

Duplicate the source bucket to the destination bucket. If `verify_md5` is true then verify md5s match. If the md5 is not availabe compute it and verify it matches.
## Deploy

Supervisord is being used to deploy. Exact configuration is coming.
## LICENSE

See the [LICENSE file](/LICENSE).

jmatt committed:
README and initial template.

	modified:   .gitignore
	new file:   README.md
	new file:   requirements.txt
	new file:   s3s3/__init__.py

jmatt committed:
Initial functioning version.

	* Configurable source and destination S3 services.
	* Pluggable notification method. Though only redis is implemented.
	* API upload method will download to a tempfile then uploads
	  to multiple destination keys.

	modified:   .gitignore
	modified:   requirements.txt
	modified:   s3s3/__init__.py
	new file:   s3s3/api.py
	new file:   s3s3/client.py
	new file:   s3s3/config.py
	new file:   s3s3/pubsub/__init__.py
	new file:   s3s3/pubsub/redis.py
	new file:   s3s3/s3s3.ini.dist

jmatt committed:
Remove extra lines in .dist.

	modified:   s3s3/s3s3.ini.dist

jmatt committed:
Pass all destination keys so s3 key only needs to be fetched once.

	modified:   s3s3/api.py
	modified:   s3s3/client.py
	modified:   s3s3/config.py

jmatt committed:
Fix typo in python-daemon dependency.

	modified:   requirements.txt

jmatt committed:
Add redis==2.10.3 to requirements.

	modified:   requirements.txt

jmatt committed:
Fix typo. dest_key => dest_keys.

	modified:   s3s3/api.py

jmatt committed:
Don't use python 3.5 destructuring since server is 3.4.x.

	modified:   s3s3/config.py

jmatt committed:
Fix syntax error when calling connection_fixup.

	modified:   s3s3/config.py

jmatt committed:
Remove print in s3s3.config module.

	modified:   s3s3/config.py

jmatt committed:
After upload, add backup verification key to redis.

	modified:   s3s3/api.py

jmatt committed:
Flush to the NamedTemporaryFile before using it.

	modified:   s3s3/api.py

jmatt committed:
decode (utf-8) s3_key from bytes to str.

	modified:   s3s3/client.py

jmatt committed:
Typo in s3s3.api.upload.

	modified:   s3s3/api.py

jmatt committed:
Add duplicate_bucket function to api.

    * Ceph, and possibly other s3 clones, are terrible and don't have
      an md5 stored. So fetch the object temporarily to force boto to
      calculate the md5.

	modified:   s3s3/api.py

jmatt committed:
Import boto.s3.key.Key.

	modified:   s3s3/api.py

jmatt committed:
Fix variable reference. Use source_key.key instead of s3_key.

	modified:   s3s3/api.py

jmatt committed:
Fix variable name. key not dest_key.

	modified:   s3s3/api.py

jmatt committed:
Use better variable name conn_dicts.

	modified:   s3s3/config.py

jmatt committed:
Add verify_md5 parameter to duplicate_bucket.

    * Defaults to False.

	modified:   s3s3/api.py

jmatt committed:
MVP version of s3s3.

    * Improve documentation.
    * Add verify_md5.
    * Improve config and client structure.

	modified:   s3s3/__init__.py
	modified:   s3s3/api.py
	modified:   s3s3/client.py
	modified:   s3s3/config.py
	modified:   s3s3/pubsub/__init__.py
	modified:   s3s3/pubsub/redis.py
	modified:   s3s3/s3s3.ini.dist

jmatt committed:
MVP version of bucket and listen scripts.

	new file:   s3s3/scripts/__init__.py
	new file:   s3s3/scripts/bucket.py
	new file:   s3s3/scripts/listen.py

jmatt committed:
Add setup.py and version module.

	new file:   s3s3/version.py
	new file:   setup.py

jmatt committed:
Add MANIFEST.in file.

	new file:   MANIFEST.in

jmatt committed:
Limit python classifiers to Python 3.4 and 3.5.

	modified:   setup.py

jmatt committed:
Increment version to 0.1.1.

	  * Fix config module so it won't raise an exception when
            initializing with no ini file.

	modified:   s3s3/config.py
	modified:   s3s3/version.py

jmatt committed:
Increment version to 0.1.2.

	  * Fix bug where specific import of variables caused a copy of
	    the variable on import. Instead imported the module which
	    keeps variables 'by reference'.

	modified:   s3s3/client.py
	modified:   s3s3/version.py

jmatt committed:
Remove reference to an older repo. /wave @craigjb

	modified:   setup.py

jmatt committed:
Increment version to 0.1.3.

	modified:   s3s3/version.py

jmatt committed:
Increment version.

	  * Remove dependency.

	modified:   requirements.txt
	modified:   s3s3/version.py

jmatt committed:
Update summary description.

	modified:   README.md

jmatt committed:
Improve README.md documentation and organization.

	modified:   README.md

jmatt committed:
Remove triple quote in Client section.

	modified:   README.md

jmatt committed:
Add supervisord conf for s3s3listen and s3s3bucket.

	new file:   extras/s3s3.conf

jmatt committed:
Increment version to 0.1.5.

	  * Add supervisord configuration.
	  * Add echo_s3s3_supervisord_conf command.
	  * Add supervisord file to setup.py as a data_file.

	new file:   s3s3/scripts/echo_s3s3_supervisord_conf.py
	modified:   s3s3/version.py
	modified:   setup.py

jmatt committed:
Add deploy documentation.

	modified:   README.md

jhoblitt commented on s3s3/api.py:
```diff
@@ -0,0 +1,114 @@
+            'The source_key and dest_keys parameters are required.')
+    with tempfile.NamedTemporaryFile() as data:
+        source_key.get_contents_to_file(data)
+        data.file.flush()
```
It's unfortunate that a file descriptor can't be passed here but I agree that using temp files is probably "safer" than passing a string.

jmatt replied:
NamedTemporaryFiles are deleted on close. So the biggest concern, unrecovered resources, seems to be handled appropriately. The NameTemporaryFile scheme allows for `mode='w+b'` which allows them to be opened multiple times yet still be reliably deleted.

If Ceph (and Swift) were more S3 compliant then we could just use the current libraries and everything would be significantly simplified. We could just stream from source to destination.

jhoblitt replied:
My concern was purely about the expense of having to flush an inode that will be immediately unlinked.  There's no away around that with the boto API without passing strings, which risk memory exhaustion.  I don't expect enough repo churn that this will be a significant issue in practice.

jhoblitt commented on s3s3/api.py:
```diff
@@ -0,0 +1,114 @@
+        source_key.get_contents_to_file(data)
+        data.file.flush()
+        for dest_key in dest_keys:
+            dest_key.set_contents_from_filename(data.name)
```
I think we want to be able to prevent existing objects from being overwritten.

jmatt replied:
I'm OK with this change.

The reason I implemented this way is because by the time upload is called the intent is to upload to the destination. If you want to ensure that they have the same data then verify_md5 should be used.

jmatt replied:
`upload` should only be called when `not dest_key.exists()`. See: [on_notify](https://github.com/jmatt/s3s3/blob/master/s3s3/client.py#L26) and [duplicate_bucket](https://github.com/jmatt/s3s3/blob/master/s3s3/api.py#L72).

I can still add it though since it could happen if someone is using the s3s3.api.upload directly.

jhoblitt replied:
Since it shouldn't cause additional network I/O, my vote would be to add an option to control overwriting that defaults to false.

jhoblitt commented on s3s3/api.py:
```diff
@@ -0,0 +1,114 @@
+        if verify_md5:
+            _update_md5([source_key, dest_key])
+            if source_key.md5 != dest_key.md5:
+                upload(source_key, [dest_key])
```
As with my previous comment, I don't think we want to do this unconditionally.

jmatt replied:
Same reasoning here as with s3s3.api.upload. Whatever the decision it's a simple change.

jhoblitt replied:
My preference is for a tinfoil hat...

jhoblitt commented on s3s3/api.py:
```diff
@@ -0,0 +1,114 @@
+    for key in keys:
+        if not key.md5:
+            with tempfile.NamedTemporaryFile() as data:
+                key.get_contents_to_file(data)
```
This has a side effect of causing the md5 to be recalculated?  https://github.com/boto/boto/blob/d9e5cfe900e1a58717e393c76a6e3580305f217a/boto/file/key.py#L134-L161

jmatt replied:
The code for [boto.s3.key.Key.md5 (boto==2.38.0)](https://github.com/boto/boto/blob/2.38.0/boto/s3/key.py#L167-L177).

Specifically the [code that updates md5 (and other digesters) is in `_get_file_internal`](https://github.com/boto/boto/blob/2.38.0/boto/s3/key.py#L1554-L1555).

The problem is as seen from the ipython:

```
In [23]: key._get_remote_metadata()
Out[23]: {'content-type': 'application/octet-stream'}
```

Ceph doesn't serve up the content-md5 in it's response.

```
curl -i https://s3.lsst.codes/lsst-git-lfs/data/f6b0597622934ba80fabc1d1c6afb83004215e9fc00432350444c18082030d2f -o f6b0597622934ba80fabc1d1c6afb83004215e9fc00432350444c18082030d2f
http://docs.ceph.com/docs/master/radosgw/s3/objectops/
```

which has these response headers.

```
HTTP/1.1 200 OK
Date: Thu, 29 Oct 2015 04:02:46 GMT
Server: Apache/2.4.7 (Ubuntu)
Accept-Ranges: bytes
Strict-Transport-Security: max-age=63072000; includeSubDomains
X-Frame-Options: SAMEORIGIN
Content-Length: 98150412
Last-Modified: Tue, 20 Oct 2015 17:09:48 GMT
ETag: "8cbd36bc39ab5e0c4ad0e8469534c5e5"
Public-Key-Pins: pin-sha256="<Subject Public Key Information (SPKI)>"; max-age=2592000; includeSubDomains
Content-Type: application/octet-stream
```

More docs about what Ceph implements.

http://docs.ceph.com/docs/master/radosgw/s3/objectops/
http://docs.ceph.com/docs/master/dev/radosgw/s3_compliance/

jhoblitt commented:
This looks reasonable but I'll note that a project of this complexity generally should have unit tests.

jmatt committed:
Add echo_s3s3_ini_template command.

    * Move s3s3.s3s3.ini.dist to extras/s3s3.ini.dist.
    * Add script to echo_s3s3_ini_template
    * Add entrypoint to setup.py.
    * Add `pubsub` section to s3s3.ini.dist.

	renamed:    s3s3/s3s3.ini.dist -> extras/s3s3.ini.dist
	new file:   s3s3/scripts/echo_s3s3_ini_template.py
	modified:   setup.py

jmatt committed:
Add commented `cp` s3s3.ini command to deploy section.

	modified:   README.md

jmatt committed:
Increment version to 0.1.6.

	  * Add a log module.
	  * Try to use /var/log/s3s3/s3s3.log. Otherwise use /tmp/s3s3.log.
	  * Use the log module's logger everywhere.
	  * PEP8 Compliance.

	modified:   README.md
	modified:   s3s3/__init__.py
	modified:   s3s3/api.py
	modified:   s3s3/client.py
	modified:   s3s3/config.py
	new file:   s3s3/log.py
	modified:   s3s3/pubsub/__init__.py
	modified:   s3s3/pubsub/redis.py
	modified:   s3s3/scripts/bucket.py
	modified:   s3s3/scripts/echo_s3s3_ini_template.py
	modified:   s3s3/scripts/listen.py
	modified:   s3s3/version.py
	modified:   setup.py

jmatt committed:
Increment version to 0.1.8.

	  * Use pkg_resource to retrieve data files in echo_* scripts.
	  * Support sdist and bdist_wheel resource locations.
	  * Add extras/*.dist and extras/*.conf to MANIFEST.in.

	modified:   MANIFEST.in
	modified:   s3s3/scripts/echo_s3s3_ini_template.py
	modified:   s3s3/scripts/echo_s3s3_supervisord_conf.py
	modified:   s3s3/version.py
	modified:   setup.py

jmatt committed:
Add echo_s3s3_ini_template to deploy section of README.md.

	modified:   README.md

jmatt commented:
I agree that the library needs unit tests going forward. I'll add it to my low priority TODO's for this cycle.

jmatt committed:
Add Command Line section to README.md.

    * Change ordering so Client comes before the Command Line section.

	modified:   README.md

jmatt committed:
Add `force` kwarg for s3s3.api.upload.

    * Change order operations so we fail early and do any real
      IO as late as possible.
    * Thanks @chrisbrue for `force` idea.

	modified:   s3s3/api.py

jmatt committed:
Increment version to 0.1.9.

          * Add dest_key.exists() check in upload.
	  * Change order of operations to fail early and
	    do IO as late as possible.

	modified:   s3s3/version.py

jonathansick committed:
Convert README to reStructuredText

jmatt committed:
Merge pull request #1 from jonathansick/rst_docs

Convert README to reStructuredText

jmatt commented:
Approved by @jhoblitt. Merge incoming.
